src/bucket/BucketManagerImpl.cpp

```diff
@@ -907,90 +922,135 @@ BucketManagerImpl::maybeSetIndex(std::shared_ptr<Bucket> b,
 }
 
 void
-BucketManagerImpl::scanForEviction(AbstractLedgerTxn& ltx, uint32_t ledgerSeq)
+BucketManagerImpl::scanForEvictionLegacySQL(AbstractLedgerTxn& ltx,
+                                            uint32_t ledgerSeq)
 {
     ZoneScoped;
-    if (protocolVersionStartsFrom(ltx.getHeader().ledgerVersion,
-                                  SOROBAN_PROTOCOL_VERSION))
-    {
-        mBucketList->scanForEviction(mApp, ltx, ledgerSeq,
-                                     mBucketListEvictionCounters);
-    }
+    releaseAssert(protocolVersionStartsFrom(ltx.getHeader().ledgerVersion,
+                                            SOROBAN_PROTOCOL_VERSION));
+    mBucketList->scanForEvictionLegacySQL(
+        mApp, ltx, ledgerSeq, mBucketListEvictionCounters, mEvictionStatistics);
 }
 
-medida::Timer&
-BucketManagerImpl::recordBulkLoadMetrics(std::string const& label,
-                                         size_t numEntries) const
+void
+BucketManagerImpl::startBackgroundEvictionScan(uint32_t ledgerSeq)
+{
+    releaseAssert(mApp.getConfig().isUsingBucketListDB());
+    releaseAssert(!mEvictionFuture.valid());
+
+    auto searchableBL = getSearchableBucketListSnapshot();
+    auto const& cfg = mApp.getLedgerManager().getSorobanNetworkConfig();
+    auto const& sas = cfg.stateArchivalSettings();
+
+    using task_t = std::packaged_task<EvictionResult()>;
+    auto task = std::make_shared<task_t>(
+        [bl = move(searchableBL), iter = cfg.evictionIterator(), ledgerSeq, sas,
+         &counters = mBucketListEvictionCounters,
+         &stats = mEvictionStatistics] {
+            return bl->scanForEviction(ledgerSeq, counters, iter, stats, sas);
+        });
+
+    mEvictionFuture = task->get_future();
+    mApp.postOnBackgroundThread(bind(&task_t::operator(), task),
+                                "SearchableBucketListSnapshot: eviction scan");
+}
+
+void
+BucketManagerImpl::resolveBackgroundEvictionScan(
+    AbstractLedgerTxn& ltx, uint32_t ledgerSeq,
+    LedgerKeySet const& modifiedKeys)
 {
-    if (numEntries != 0)
+    ZoneScoped;
+
+    if (!mEvictionFuture.valid())
     {
-        mBucketListDBBulkLoadMeter.Mark(numEntries);
+        startBackgroundEvictionScan(ledgerSeq);
     }
 
-    auto iter = mBucketListDBBulkTimers.find(label);
-    if (iter == mBucketListDBBulkTimers.end())
+    mEvictionFuture.wait();
+    auto evictionCandidates = mEvictionFuture.get();
+
+    auto const& networkConfig =
+        mApp.getLedgerManager().getSorobanNetworkConfig();
+
+    // If eviction related settings changed during the ledger, we have to
+    // restart the scan
+    if (!evictionCandidates.isValid(ledgerSeq,
+                                    networkConfig.stateArchivalSettings()))
     {
-        auto& metric =
-            mApp.getMetrics().NewTimer({"bucketlistDB", "bulk", label});
-        iter = mBucketListDBBulkTimers.emplace(label, metric).first;
+        startBackgroundEvictionScan(ledgerSeq);
```

Comment:
While `scanForEvictionLegacySQL` and the background eviction scan should behave identically, there implementations are fairly different. The goal is to have background scanning by default and eventually deprecate the synchronous scan. For this reason, I think it's safer to only call variant of the scan, and of the two I default to the background scan when it's enabled. Personally, I think it's less of a footgun only calling into one implementation instead of mixing and matching 2.